src/guardian/keys-validation/keys-validation.service.ts

```diff
@@ -27,29 +43,54 @@ export class KeysValidationService {
   ): Promise<{ key: string; depositSignature: string }[]> {
     const forkVersion: Uint8Array = await this.forkVersion();
 
-    const validatedKeys: [
-      {
-        key: string;
-        depositSignature: string;
-        used: boolean;
-      },
-      boolean,
-    ][] = await this.keyValidator.validateKeys(
-      vettedKeys.map((key) => ({
-        key: key.key,
-        depositSignature: key.depositSignature,
-        used: false,
-        withdrawalCredentials: bufferFromHexString(withdrawalCredentials),
-        genesisForkVersion: Buffer.from(forkVersion.buffer),
-      })),
-    );
+    const keysForValidation = vettedKeys
+      .map((key) =>
+        this.prepareKeyForValidation(key, withdrawalCredentials, forkVersion),
+      )
+      .filter((key) => key !== undefined) as DepositData[];
+
+    const validatedKeys: [Key & DepositData, boolean][] =
+      await this.keyValidator.validateKeys(keysForValidation);
+
+    this.updateCacheWithValidationResults(validatedKeys);
 
-    return validatedKeys
-      .filter(([, result]) => !result)
+    // this list will not include invalid keys from cache
+    const invalidKeysFromCurrentValidation = validatedKeys
+      .filter(([, isValid]) => !isValid)
       .map(([key]) => ({
         key: key.key,
         depositSignature: key.depositSignature,
       }));
+
+    return this.mergeInvalidKeys(vettedKeys, invalidKeysFromCurrentValidation);
+  }
+
+  prepareKeyForValidation(
+    key: RegistryKey,
+    withdrawalCredentials: string,
+    forkVersion: Uint8Array,
+  ) {
+    const cachedEntry = this.keysCache.get(key.key);
+
+    // key wasn't in cache or signature was changed
+    if (cachedEntry && cachedEntry.signature == key.depositSignature) {
```

Comment:
prepare function can return null. Rename this function to filter and use it like filter

Reply:
fixed